File: client/lib/modules/documents/steam_data.dart
class SteamData {
  final String name;
  final int steamAppid;
  final String detailedDescription;
  final String shortDescription;
  final String aboutTheGame;

  final String? headerImage;
  final String? backgroundImage;

  final List<Screenshot> screenshots;
  final List<Movie> movies;

  const SteamData({
    required this.name,
    required this.steamAppid,
    required this.detailedDescription,
    required this.shortDescription,
    required this.aboutTheGame,
    this.headerImage = '',
    this.backgroundImage = '',
    this.screenshots = const [],
    this.movies = const [],
  });

  SteamData.fromJson(Map<String, dynamic> json)
      : this(
          name: json['name']!,
          steamAppid: json['steam_appid']!,
          detailedDescription: json['detailed_description']!,
          shortDescription: json['short_description']!,
          aboutTheGame: json['about_the_game']!,
          headerImage: json['header_image'],
          backgroundImage: json['background_raw'],
          screenshots: json.containsKey('screenshots')
              ? [
                  for (final screenshot in json['screenshots'])
                    Screenshot.fromJson(screenshot),
                ]
              : [],
          movies: json.containsKey('movies')
              ? [
                  for (final movie in json['movies']) Movie.fromJson(movie),
                ]
              : [],
        );

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'steam_appid': steamAppid,
      'detailed_description': detailedDescription,
      'short_description': shortDescription,
      'about_the_game': aboutTheGame,
      if (headerImage != null) 'header_image': headerImage,
      if (backgroundImage != null) 'background_raw': backgroundImage,
      if (screenshots.isNotEmpty)
        'screenshots': [
          for (final screenshot in screenshots) screenshot.toJson(),
        ],
      if (movies.isNotEmpty)
        'movies': [
          for (final movie in movies) movie.toJson(),
        ],
    };
  }
}

class Screenshot {
  final int id;
  final String pathThumbnail;
  final String pathFull;

  const Screenshot({
    required this.id,
    required this.pathThumbnail,
    required this.pathFull,
  });

  Screenshot.fromJson(Map<String, dynamic> json)
      : this(
          id: json['id'],
          pathThumbnail: json['path_thumbnail'],
          pathFull: json['path_full'],
        );

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'path_thumbnail': pathThumbnail,
      'path_full': pathFull,
    };
  }
}

class Movie {
  final int id;
  final String name;
  final String thumbnail;
  final WebM webm;

  const Movie({
    required this.id,
    required this.name,
    required this.thumbnail,
    required this.webm,
  });

  Movie.fromJson(Map<String, dynamic> json)
      : this(
          id: json['id'],
          name: json['name'],
          thumbnail: json['thumbnail'],
          webm: WebM.fromJson(json['webm']),
        );

  Map<String, dynamic> toJson() {
    return {
      'image_id': id,
      'height': name,
      'width': thumbnail,
    };
  }
}

class WebM {
  final String max;

  const WebM({
    required this.max,
  });

  WebM.fromJson(Map<String, dynamic> json)
      : this(
          max: json['max'],
        );

  Map<String, dynamic> toJson() {
    return {
      'max': max,
    };
  }
}

File: server/src/documents/steam_data.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct SteamData {
    pub name: String,
    pub steam_appid: u64,
    pub detailed_description: String,
    pub short_description: String,
    pub about_the_game: String,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub header_image: Option<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub background_raw: Option<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub developers: Vec<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub publishers: Vec<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub dlc: Vec<u64>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub genres: Vec<Genre>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub screenshots: Vec<Screenshot>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub movies: Vec<Movie>,
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct Genre {
    pub id: String,
    pub description: String,
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct Screenshot {
    pub id: u64,
    pub path_thumbnail: String,
    pub path_full: String,
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct Movie {
    pub id: u64,
    pub name: String,
    pub thumbnail: String,
    pub webm: WebM,
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct WebM {
    pub max: String,
}
